pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,450 @@
+        match timeline.evict_layer(&layer.file_name.file_name()).await {
+            Err(e) => {
+                warn!(%tenant_id, %timeline_id, "failed to evict layer: {:#}", e);
+            }
+            Ok(None) | Ok(Some(false)) => {
+                info!(%tenant_id, %timeline_id, "layer does not exist, likely it was deleted");
+                // XXX should we account for this in usage_expected.avail_bytes ?
+            }
+            Ok(Some(true)) => {
+                usage_expected.set_avail_bytes(usage_expected.avail_bytes() + layer.file_size);
+            }
+        }
+        if !usage_expected.with_pressure(task_config).has_pressure() {
+            break;
+        }
+    }
+
+    // Verify with statvfs whether we made any real progress
+    let usage_post = filesystem_level_usage::get(tenants_dir_fd)
+        // It's quite unlikely to hit the error here. Keep the code simple and bail out.
+        .context("get filesystem-level disk usage after evictions")?
+        .with_pressure(task_config);
+
+    debug!(?usage_post, "disk usage");
+
+    Ok(IterationOutcome::Finished(IterationOutcomeFinished {
+        before: usage_pre,
+        after_actual: usage_post,
+        after_expected: usage_expected,
+    }))
+}
+
+mod filesystem_level_usage {
+    use anyhow::Context;
+    use nix::{
+        dir::Dir,
+        sys::statvfs::{self, Statvfs},
+    };
+    use sync_wrapper::SyncWrapper;
+
+    use super::DiskUsageEvictionTaskConfig;
+
+    // The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+    // We use the Debug impl for logging, so, it's allright.
+    #[derive(Debug, Clone)]
+    #[allow(dead_code)]
+    pub struct Usage {
+        total_bytes: u64,
+        avail_bytes: u64,
+        usage_pct_cache: Option<u64>,
+    }
+
+    pub type Pressure = [(&'static str, bool); 2];
+
+    // The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+    // We use the Debug impl for logging, so, it's allright.
+    #[derive(Debug)]
+    #[allow(dead_code)]
+    pub struct UsageWithPressure {
+        pub usage: Usage,
+        pub pressure: Pressure,
+    }
+
+    impl UsageWithPressure {
+        pub fn has_pressure(&self) -> bool {
+            self.pressure.iter().any(|(_, has_pressure)| *has_pressure)
+        }
+    }
+
+    impl Usage {
+        pub fn avail_bytes(&self) -> u64 {
+            self.avail_bytes
+        }
+
+        pub fn set_avail_bytes(&mut self, avail_bytes: u64) {
+            self.avail_bytes = avail_bytes;
+            self.usage_pct_cache = None;
+        }
+
+        pub fn usage_pct(&mut self) -> u64 {
+            self.usage_pct_cache.unwrap_or_else(|| {
+                let usage_pct = (100.0
+                    * (1.0 - ((self.avail_bytes as f64) / (self.total_bytes as f64))))
+                    as u64;
+                self.usage_pct_cache = Some(usage_pct);
+                usage_pct
+            })
```

Comment:
Just noticed: This will not update the cached value.


```suggestion
            if self.usage_pct_cache.is_none() {
              self.usage_pct_cache = Some({
                let usage_pct = (100.0
                    * (1.0 - ((self.avail_bytes as f64) / (self.total_bytes as f64))))
                    as u64;
                self.usage_pct_cache = Some(usage_pct);
                usage_pct
              });
            }
            
            self.usage_pct_cache.unwrap()
```

Reply:
Related: debug fmt TODO, but we can resolve this. I did notice in logs that it has turned `Some` in some cases.